twister: pytest: Wrap iter_stdout

gchwier opened the pull request:
Added waiting for prompt instead of calling `time.sleep`, to fix issue with longer setting up platforms (e.g. with TF-M enabled).
Wrapped `iter_stdout` generator, to avoid issues when generator is created and called multiple times in test scenarios.

To run and test one can run:
`./scripts/twister -vv -T samples/subsys/testsuite/pytest/shell -p qemu_x86 -p qemu_cortex_m3 -p native_posix`
or on DK:
`./scripts/twister -vv -T samples/subsys/testsuite/pytest/shell -p nrf9160dk_nrf9160_ns  --device-testing --device-serial /dev/ttyACM0`

Fixes: #58747

gchwier committed:
twister: pytest: Wrap iter_stdout

Added waiting for prompt instead of calling time.sleep,
to fix issue with longer setting up platforms with TF-M
enabled. Wrapped iter_stdout generator, to avoid issues
when generator is created and called multiple times
in test scenarios.
Fixes: #58747

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

PerMac commented:
Observation: the PR introducing pytest integration required retries on several qemus (e.g. https://github.com/zephyrproject-rtos/zephyr/actions/runs/5091718306/jobs/9153141110#step:13:834). This PR didn't need any retry for pytest shell test. We also observed that our hw platform using TF-m were always failing, due to not enough time for the app to boot and with this fix they no longer fail.

pdgendt commented on samples/subsys/testsuite/pytest/shell/pytest/test_shell.py:
```diff
@@ -22,15 +21,25 @@ def wait_for_message(iter_stdout, message, timeout=60):
+    time_started = time.time()
+    while True:
+        dut.write(b'\n')
+        for line in dut.iter_stdout:
```
How is this guaranteed that it won't block forever?

gchwier replied:
It shouldn't block. `iter_stdout` calls `readline` method, that is blocking, but in device adapters there is implementation to avoid blocking forever: in `qemu_adapter` and `simulator_adapter` (for native_posix) are threads and queues used to set timeout. In `hardware_adapter` the timeout is passed during Serial initialization.

gchwier replied:
Unfortunately, when using Serial (for hardware) the `readline` is blocking. We must change this helper method. Thanks @pdgendt for comment

PerMac commented:
I temporarily added DNM since there is still a possibility that an error during a test on HW can block the further execution. We are looking for the fix.

gopiotr committed:
scripts: twister: add timeout for pytest process

Add protection timeout for pytest subprocess, to avoid situation of
suspending whole Twister in case of internal pytest test problem.

Co-authored-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>
Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

gchwier commented:
@pdgendt @PerMac 
A protection timeout for pytest subprocess was added, to avoid situation of suspending whole Twister in case of internal pytest test problem (such as calling blocking `readline` method on serial). 
We have listed in known limitation in #58288 that `iter_stdout` needs some further investigation (some improvements to read bytes instead of lines, add timeout etc.), so it will be probable changed in next iterations.
